model-server/src/main/kotlin/org/modelix/model/server/store/IStoreClient.kt

```diff
@@ -20,29 +20,164 @@ import kotlinx.coroutines.coroutineScope
 import kotlinx.coroutines.launch
 import kotlinx.coroutines.withContext
 import kotlinx.coroutines.withTimeoutOrNull
-import org.modelix.model.IKeyListener
+import org.modelix.kotlin.utils.ContextValue
+import org.modelix.model.IGenericKeyListener
+import org.modelix.model.lazy.RepositoryId
+import org.modelix.model.persistent.HashUtil
 import java.io.File
 import java.io.IOException
 import kotlin.time.Duration.Companion.seconds
 
-interface IStoreClient : AutoCloseable {
-    operator fun get(key: String): String?
-    fun getAll(keys: List<String>): List<String?>
-    fun getAll(keys: Set<String>): Map<String, String?>
-    fun getAll(): Map<String, String?>
-    fun put(key: String, value: String?, silent: Boolean = false)
-    fun putAll(entries: Map<String, String?>, silent: Boolean = false)
-    fun listen(key: String, listener: IKeyListener)
-    fun removeListener(key: String, listener: IKeyListener)
-    fun generateId(key: String): Long
+/**
+ * Stores objects of a hash tree. Valid entries are required to meet the condition `key == HashUtil.sha256(value)`.
+ */
+interface IObjectStore {
+    fun getAll(key: Set<String>): Map<String, String>
+    fun putAll(entries: Map<String, String?>)
+}
+
+interface IGenericStoreClient<KeyT> : AutoCloseable {
+    operator fun get(key: KeyT): String? = getAll(listOf(key)).first()
+    fun getAll(keys: List<KeyT>): List<String?> {
+        val entries = getAll(keys.toSet())
+        return keys.map { entries[it] }
+    }
+    fun getAll(keys: Set<KeyT>): Map<KeyT, String?>
+    fun getAll(): Map<KeyT, String?>
+    fun put(key: KeyT, value: String?, silent: Boolean = false) = putAll(mapOf(key to value))
+    fun putAll(entries: Map<KeyT, String?>, silent: Boolean = false)
+    fun listen(key: KeyT, listener: IGenericKeyListener<KeyT>)
+    fun removeListener(key: KeyT, listener: IGenericKeyListener<KeyT>)
+    fun generateId(key: KeyT): Long
     fun <T> runTransaction(body: () -> T): T
 }
 
+interface IStoreClient : IGenericStoreClient<String>
+typealias IIsolatedStore = IGenericStoreClient<ObjectInRepository>
+
+class RepositoryScopedStoreClient(private val repositoryId: RepositoryId, client: IGenericStoreClient<ObjectInRepository>) : StoreClientAdapter(client) {
+    override fun getRepositoryId() = repositoryId
+}
+
+class GlobalStoreClient(client: IGenericStoreClient<ObjectInRepository>) : StoreClientAdapter(client) {
+    override fun getRepositoryId() = null
+}
+
+class ContextScopedStoreClient(client: IGenericStoreClient<ObjectInRepository>) : StoreClientAdapter(client) {
```

Comment:
thought: I see the use of a context scope as accidental complexity.
I also see why it helps. It reduces the need for code rewrite to introduce this feature.

In the future, we could get rid of it by letting by breaking up the `IgniteStoreClient` to into two parts. One for model data and one for metadata (e.g., repositories, branches, cliendIds, serverIds etc.) and use them everywhere without mixing them together.

I am not too concerned about it because refactoring to get rid of `ContextScopedStoreClient` is possible without changing the data schema.

Reply:
I think splitting the StoreClient is a more complex solution. That moves responsibilities about transaction management out of the StoreClient to users of it.

Reply:
Btw, looking at the DDD community, transaction management is usually understood as something that has to happen on the highest application level as it impacts the domain-specific consistency guarantees you want or need to give. That's also the reason why in Spring or Quarkus, usually either the REST controllers or the called application service method (assuming each request is served by a signle application service method) is marked as `@Transactional`.

Reply:
We already documented to do it that way in https://issues.modelix.org/articles/MODELIX-A-91/0010-Dispatch-blocking-code-directly-from-request-handlers.

Reply:
> I think splitting the StoreClient is a more complex solution. That moves responsibilities about transaction management out of the StoreClient to users of it.

Good point.